src/core/utils/resolveUnderlying.ts

```diff
@@ -0,0 +1,151 @@
+import { Protocol } from '../../adapters/protocols'
+import { Erc20Metadata } from '../../types/erc20Metadata'
+import { IProtocolAdapter } from '../../types/IProtocolAdapter'
+import {
+  AdapterMissingError,
+  ProtocolSmartContractNotDeployedAtRequestedBlockNumberError,
+} from '../errors/errors'
+import { logger } from './logger'
+
+type Token = Erc20Metadata & {
+  tokens?: Token[]
+  priceRaw?: bigint
+}
+
+export async function resolveUnderlyings(
+  adapter: IProtocolAdapter,
+  blockNumber: number | undefined,
+  tokens: Token[],
+  updateUnderlyingToken: (
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    underlyingToken: any,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    protocolToken: any,
```

Comment:
There has to be a way of sorting these `any` with generics, but I'm hitting a brick wall right now, so I need to give it a few hours to see if the penny drops.